akd_core/src/utils.rs

```diff
@@ -33,37 +42,140 @@ pub fn get_marker_version_log2(version: u64) -> u64 {
+/// MARKER_VERSION_SKIPLIST that is greater than start_version.
+/// 2. Include all elements of MARKER_VERSION_SKIPLIST that are between start_version and epoch.
+/// 3. Include at most a log_2(start_version) number of versions between start_version and the
+/// smallest power of 2 greater than start_version, determined as follows: For each bit position i
+/// in start_version, if the bit is 0, include the value of start_version with the ith bit set
+/// to 1 and followed by trailing zeros.
+///
+/// As a concrete example, if start_version = 85, the future marker versions would be
+/// [86, 88, 96, 128, 256, 65536, 2^32] (potentially truncated depending on if any of these
+/// numbers exceed epoch).
+///
+/// Since:
+/// 01010101 => 85
+/// 01010110 => 86
+/// 01011000 => 88
+/// 01100000 => 96
+/// 10000000 => 128
+///
+/// And the remainder of the list comes from MARKER_VERSION_SKIPLIST.
+///
+/// Note that the past marker versions do not contain start_version, as this would be redundant
+/// in the history proof (since membership is already checked for start_version).
 pub fn get_marker_versions(
     start_version: u64,
     end_version: u64,
     epoch: u64,
 ) -> (PastMarkerVersions, FutureMarkerVersions) {
     // Compute past marker versions
     let mut past_marker_versions: Vec<u64> = Vec::new();
-    let start_marker = 1 << get_marker_version_log2(start_version);
-    if start_marker < start_version {
-        past_marker_versions.push(start_marker);
+
+    let skiplist_past_index: usize =
+        find_max_index_in_array(start_version, &MARKER_VERSION_SKIPLIST);
+    if MARKER_VERSION_SKIPLIST[skiplist_past_index] != start_version {
+        past_marker_versions.push(MARKER_VERSION_SKIPLIST[skiplist_past_index]);
+    }
+    let start_version_log2 = 1 << get_marker_version_log2(start_version);
+    if start_version_log2 != start_version
+        && (past_marker_versions.is_empty()
+            || start_version_log2 != past_marker_versions[past_marker_versions.len() - 1])
+    {
+        past_marker_versions.push(start_version_log2);
+    }
+
+    let start_version_length = get_bit_length(start_version);
+    for i in (0..start_version_length).rev() {
+        let shift = 1 << i;
+        // Check if the bit of start_version at position i is 1
+        if start_version & shift != 0 {
+            let shift_mask = (shift - 1) | shift;
+            let past = start_version & !shift_mask;
+            if past != 0
+                && (past_marker_versions.is_empty()
+                    || past != past_marker_versions[past_marker_versions.len() - 1])
+            {
+                past_marker_versions.push(past);
+            }
+        }
     }
 
     // Compute future marker versions
+    let mut future_marker_versions: Vec<u64> = Vec::new();
+
+    let end_version_length = get_bit_length(end_version);
+    let mut future: u64 = end_version;
```

Comment:
Very small nit pick: shall we consider renaming this to something that isn't easily mistaken for a future (in the async sense).

Reply:
Renamed to future_version (and the other variable from past -> past_version)